### PR TITLE
ENH helper for non diagonal covariance in simulated

### PR DESCRIPTION
**Pull request by mathurinm**
Addressing #21, but it takes a long time (> 3 mins) to generate the data now (my guess is that an EVD of correlation matrix is performed, the later being 10k x 10k in some cases)

### CODE REVIEW

File: benchmarks/lasso/datasets/simulated.py [hunk 1]
```diff
@@ -12,22 +12,26 @@ class Dataset(BaseDataset):
     parameters = {
         'n_samples, n_features': [
             (100, 5000),
-            (100, 10000)]
+            (100, 10000)],
     }
 
-    def __init__(self, n_samples=10, n_features=50, random_state=27):
+    def __init__(self, n_samples=10, n_features=50, random_state=27, corr=0.5):
```
Comment (agramfort) [inline benchmarks/lasso/datasets/simulated.py]: you should use corr = 0 by default

File: benchmarks/lasso/datasets/simulated.py [hunk 2]
```diff
@@ -12,22 +12,26 @@ class Dataset(BaseDataset):
...
         # Pass parameters that will be used in Dataset name
         super().__init__(n_samples=n_samples, n_features=n_features)
 
     def get_data(self):
-
+        # use Toeplitz covariance matrix:
+        idx = np.arange(self.n_features)
+        corr_mat = self.corr ** np.abs(idx[:, None] - idx)
         rng = np.random.RandomState(self.random_state)
-        X = rng.randn(self.n_samples, self.n_features)
```
Comment (agramfort) [inline benchmarks/lasso/datasets/simulated.py]: keep randn if corr == 0

**Comment by agramfort** [comment]
what is the mechanism for testing this?

**Comment by tomMoral** [comment]
This PR needs to be moved in [benchopt/benchmark_lasso](https://github.com/benchopt/benchmark_lasso).

Also, I realized today that a fast way to generate correlated matrices is to use AR processes :)
For an AR(1) of coefficient rho, the correlation matrix is the one we are looking for I guess:

```
rho = .6
res = []
for _ in range(10):
    res.append(X)
    X = rho * X + np.random.randn(p)
D = np.array(res)
D -= D.mean(axis=1, keepdims=True)
D /= D.std(axis=1, keepdims=True)
plt.matshow(D.dot(D.T) / p)
```

gives
image


WDYT?

**Comment by mathurinm** [comment]
That looks nice ! I'll give it a try in the aforementioned repos

Le lun. 28 sept. 2020 à 19:15, Thomas Moreau <notifications@github.com> a
écrit :

**Comment by tomMoral** [comment]
@mathurinm I did the change based on the AR process.
Could you review the code and tell me if you think like it is okay?

File: benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 
 from benchopt.base import BaseDataset
+from benchopt.utils.datasets.simulated import gen_correlated_data
```
Comment (mathurinm) [inline benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py:4]: maybe `make_correlated_data` as in sklearn.datasets.make_regression ?

File: benchopt/utils/checkers.py [hunk 4]
```diff
@@ -65,3 +67,24 @@ def solver_supports_dataset(solver, dataset):
         return False
 
     return True
+
+
+def check_random_state(seed):
```
Comment (mathurinm) [inline benchopt/utils/checkers.py:72]: honest question, is it better to implement ours than to import it from sklearn.utils ?

File: benchopt/utils/checkers.py [hunk 5]
```diff
@@ -65,3 +67,24 @@ def solver_supports_dataset(solver, dataset):
...
+
+    Parameters
+    ----------
+    seed : None, int or instance of RandomState
+        If seed is None, return the RandomState singleton used by np.random.
+        If seed is an int, return a new RandomState instance seeded with seed.
+        If seed is already a RandomState instance, return it.
+        Otherwise raise ValueError.
+    """
+    if seed is None or seed is np.random:
```
Comment (mathurinm) [inline benchopt/utils/checkers.py:83]: why `if seed is np.random` ?

File: benchopt/utils/datasets/simulated.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+"""Helper to generate simulated data in benchopt."""
+import numpy as np
+
+from ..checkers import check_random_state
+
+
+def gen_correlated_data(n_samples=100, n_features=50, rho=0.6,
+                        random_state=None):
+    """Generate correlated design matrix with decaying correlation rho**|i-j|.
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:9]: you could make it a verbatim string and avoid the double backslashes I think

File: benchopt/utils/datasets/simulated.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    random_state : int | RandomState instance | None (default)
+        Determines random number generation for centroid initialization and
+        random reassignment. Use an int to make the randomness deterministic.
+
+    Returns
+    -------
+    X: ndarray, shape (n_samples, n_features)
+        A design matrix with Toeplitz covariance.
+    """
+    if rho >= 1 or rho < 0:
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:43]: `if not 0 <= rho < 1:` works and is easier IMO

File: benchopt/utils/datasets/simulated.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    n_samples: int
+        Number of samples in the design matrix.
+    n_features: int
+        Number of features in the design matrix.
+    corr: float
+        Correlation $\\rho$ between successive features. The element $C_{i, j}$
+        in the correlation matrix will be $\\rho^{|i-j|}$. This parameter
+        should be selected in $[0, 1[$.
+    random_state : int | RandomState instance | None (default)
+        Determines random number generation for centroid initialization and
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:35]: maybe centroid is copy paster from somewhere else ? I don't get what it is

File: benchopt/utils/datasets/simulated.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    if rho >= 1 or rho < 0:
+        raise ValueError("The correlation `rho` should be chosen in [0, 1[.")
+    rng = check_random_state(random_state)
+
+    sigma = np.sqrt(1 - rho * rho)
+    innovation = rng.randn(n_features + 1, n_samples)
+    U = innovation[0]
+
+    res = [U.copy()]
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:51]: maybe you could preallocate ?
```
X = np.empty([n_samples, n_features])
X[:, 0] = U
for i in range(1, n_features):
    U *= rho
    U += sigma * innovation[i]
    X[:, j] = U
return X
```

Comment (mathurinm) [reply benchopt/utils/datasets/simulated.py:51]: and eventually make X fortran (not sure)

File: benchopt/utils/datasets/simulated.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    -------
+    X: ndarray, shape (n_samples, n_features)
+        A design matrix with Toeplitz covariance.
+    """
+    if rho >= 1 or rho < 0:
+        raise ValueError("The correlation `rho` should be chosen in [0, 1[.")
+    rng = check_random_state(random_state)
+
+    sigma = np.sqrt(1 - rho * rho)
+    innovation = rng.randn(n_features + 1, n_samples)
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:48]: any reason to generate them all in one go instead of calling rng for every column generation ?

File: benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py [hunk 11]
```diff
@@ -12,19 +13,25 @@ class Dataset(BaseDataset):
     parameters = {
         'n_samples, n_features': [
             (100, 5000),
-            (100, 10000)]
+            (100, 10000)],  # slow to simulate big correlated design
```
Comment (mathurinm) [inline benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py:16]: why slow ?
```
In [9]: %timeit -n 1 -r 15 gen_correlated_data(n_samples=100, n_features=10_000, rho=0.6)
91.8 ms ± 4.86 ms per loop (mean ± std. dev. of 15 runs, 1 loop each)
```

File: benchopt/utils/datasets/simulated.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    U = innovation[0]
+
+    res = [U.copy()]
+    for i in range(1, n_features):
+        U *= rho
+        U += sigma * innovation[i]
+        res.append(U.copy())
+    X = np.array(res)
+    return X.T
```
Comment (mathurinm) [inline benchopt/utils/datasets/simulated.py:57]: do we divide each column by sqrt(n_samples) ?

File: benchopt/utils/datasets/simulated.py (hunk 10 above)
Comment (tomMoral) [reply benchopt/utils/datasets/simulated.py:48]: This is faster to have only one call than multiple calls as it reduces the overhead of calling the blass I would say but I did not benchmark.

File: benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py (hunk 11 above)
Comment (tomMoral) [reply benchopt/tests/test_benchmarks/dummy_benchmark/datasets/simulated.py:16]: This is old  comment from your code, I forgot to remove.

File: benchopt/utils/datasets/simulated.py (hunk 12 above)
Comment (tomMoral) [reply benchopt/utils/datasets/simulated.py:57]: We don't in the other generation process so I would say no.
But we can if you want.

File: benchopt/utils/datasets/simulated.py (hunk 10 above)
Comment (tomMoral) [reply benchopt/utils/datasets/simulated.py]: It seems that this is not the case?!?
 
image

**Comment by mathurinm** [comment]
One last thought is that we will always need y along with X, so we generate it as well, passing `snr` and `density` default parameters:
```
coef = np.zeros(n_features)
# random Gaussian coefficients:
coef[np.random.choice(n_features, int(density * n_features), replace=False) = np.random.randn(int(density * n_features))
noise = np.random.randn(n_samples)
y = X @ coef
y += noise / norm(noise) * norm(y) / snr
return X, y
```
we may also choose to return coef for inspection

**Comment by mathurinm** [comment]
I fell this is what I have been needing for years. In sklearn make_regression I don't like the `noise` variable as it's not tangible. 

Generating y along with X seems handy, returning w_true may be useful for other metrics benchmark (spoiler)

**Comment by mathurinm** [comment]
maybe density in (0, 1) is a better parameter than nnz. No strong optinion here

File: benchopt/utils/datasets/simulated.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
+"""Helper to generate simulated data in benchopt."""
+import numpy as np
+from numpy.linalg import norm
+
+from ..checkers import check_random_state
+
+
+def make_correlated_data(n_samples=100, n_features=50, rho=0.6, snr=3,
+                         w_true=None, nnz=10, random_state=None):
```
Comment (tomMoral) [inline benchopt/utils/datasets/simulated.py:9]: Maybe change the name to `make_correlated_regression`?
Also, I would rather have a `density` parameter I think, not a big fan of having to move this if we change the number of features.

For the generation, what do you think, should we have a fixed number of coef or something that can vary a bit around the "mean" that is density?

**Review by tomMoral — COMMENTED** [review]
Just a small comment

**Comment by mathurinm** [comment]
changed to `density`. I think it's easier to explain the generatin process if the number of non zero is always density * n_features. If the user wants something more complicated, he passes his own `w_true`

**Comment by mathurinm** [comment]
Ready on my side @tomMoral

**Comment by mathurinm** [comment]
Simple test if we want to add one:
```
def test_correlated():
    X, y, w_true = make_correlated_data(snr=0)
    np.testing.assert_allclose(y, X @ w_true)

    snr = 5
    w_true = np.ones(50)
    X, y, _ = make_correlated_data(n_features=w_true.shape[0], snr=5)
    np.testing.assert_allclose(snr, norm(X @ w_true), norm(y - X @ w_true))
```

**Comment by tomMoral** [comment]
I did a small pass on tests.
Will try to merge this tomorrow.

**Comment by tomMoral** [comment]
Thx a lot @mathurinm for this marathon PR! :)

Closes #21
